coproc/docs: add example wasm + elastic search (#547)
vectorizedio/redpanda: docs/www/guide-wasm-elastic.md

# Redpanda wasm function + Elastic Search

> One of the richest wasm function case is indexing information to [ElasticSeach](https://www.elastic.co/elasticsearch) from Redpanda's records. for this we need to requirements:

Reviewer: ```suggestion
One of the richest wasm function cases is pushing to [ElasticSeach](https://www.elastic.co/elasticsearch) from a Redpanda topic. To do this, you'll need to satisfy two requirements:
```

> 1. Redpanda > v21.1.5

Reviewer 2: ```suggestion
1. Redpanda version 21.1.5 or greater. You can check it by running `rpk version`.
```

> 2. ElasticSearch service, in this case we use the [docker image](https://www.elastic.co/guide/en/elasticsearch/reference/current/docker.html), in localhost:9200

Reviewer: ```suggestion
2. ElasticSearch, in this example we use the official [docker image](https://www.elastic.co/guide/en/elasticsearch/reference/current/docker.html)
```

## Enable wasm on Redpanda

> Add a `coproc_enable: true` in redpanda.yaml

Reviewer: ```suggestion
Add this line `coproc_enable: true` to your redpanda.yaml file, under the `redpanda` section
```

Reviewer: You'll also need to say that developer mode should be set to true as well

Reviewer 2: Suggestion to the suggestion:
```diff
- Add this line `coproc_enable: true` to your redpanda.yaml file, under the `redpanda` section
+ Add `coproc_enable: true` in your redpanda.yaml file, under the `redpanda` section. You can do this manually or by running `sudo rpk redpanda config set 'redpanda.coproc_enable' true`.
```

Reviewer 2: Also, idk if it's too late, but the (ambiguous) convention I see among the current config fields is that it should be either

- `enable_xyz`
- `xyz_enabled`

https://vectorized.io/docs/advanced-configuration/

Reviewer 2: Rpk to set the developer mode ON: `sudo rpk redpanda mode dev`

Author: Done.

Author: > Also, idk if it's too late, but the (ambiguous) convention I see among the current config fields is that it should be either
> 
> * `enable_xyz`
> * `xyz_enabled`
> 
> https://vectorized.io/docs/advanced-configuration/

should we create an issue for it?

```yaml
   # Skips most of the checks performed at startup (i.e. memory, xfs)
   # not recomended for production use
   developer_mode: true
   enable_coproc: true # enable wasm
```

## Create Wasm function

> for creating a wasm function we can use `rpk` tools.

Reviewer: ```suggestion
To create a wasm function we can use `rpk` tools. It will generate a template giving us a simple build system and startup example.
```

Reviewer 2: Please show the specific `rpk wasm` command and its output.

> first, we need to set up the environment for defining and compacting

Reviewer: I would just paste the output of `tree` to show readers what to expect, then underneath explain whats important:
```
$ ~/workspace/copro: rpk wasm generate
$ ~/workspace/copro: tree
...
...
```

Author: Got it, sound good.

> first, we need to set up the environment for defining and compacting

Reviewer 2: ```suggestion
First, we need to set up the environment for defining and compacting
```

wasm function, for it we can use `rpk wasm generate`
that command creates a folder with an example wasm
definition in `current_path/wasm`, now we just to go 
`src/wasm.js`, and rename to `src/elastic.js`, remember
the wasm file name is very important, from this we create
the id for a wasm function. 

> we need the elastic search js client for this example:

Reviewer: ```suggestion
Be sure to add elasticsearch to your install dependencies
```

```bash
npm install @elastic/elasticsearch
```

> now we can edit `src/elastic.js` in order to apply our logic

Reviewer: ```suggestion
now we can edit `src/elastic.js` in order to apply our logic
```

> for ElasticSearch, you could use the next code.

Reviewer: ```suggestion
```

      if (containHeader("logService", "elastic", record)) {
        console.log("Index record value to Elastic Search")
        return client.index({
          index: "result_wasm",
          type: "redpanda_wasm",
          body: {
            /*
             index the record value into elastic search
            */
            record_message: record.value.toString()
          }
        })
      }
    })
  ).then(() => {
      /*
       return a map with recordBatch indexed to elastic Search
      */
      const result = new Map();
      result.set("elasticIndex", recordBatch)
      /* 
       Wasm processRecord function has to return a Map always,
       although that map is empty map
      */
      return result
  })
    .catch((e) => {
      /* 
        handling error by coprocessor promise and not
        coprocessor policy error 
      */
      const result = new Map();
      console.log("fail to publish record batch to Elastic Search, " +
       `record batch info`, recordBatch.header);

      /*
        saving failed recordbatch index publish 
      */
      result.set("elasticIndexFail", recordBatch)
      return result
    })
});

exports["default"] = transform;
```

> after applying the new logic, we need to compact everything into our 

Reviewer: Compact isn't the term, I think its minify? So maybe we can say "When your done creating your function, you'll need to minify(or build) it to prepare it for deployment.

Reviewer 2: `bundle` is also a good term I think: _bundle everything into a single file_ .

wasm function, for it, we can use `npm run build`, this command
generate `dist/elastic.js` file.

before deploying this wasm function, we need to create the `produce` topic

> which is where we get the records.

Reviewer: ```suggestion
The statically defined input topic that this wasm operator has defined in its metadata
```

```bash

> rpk topic create produce

Reviewer: When I put bash commands I usually prepend with `$` so it looks like this is a command you can input into the shell. I always paste the output as well, to give people who are reading this something to expect to see (on success) after the command is run

```

## Deploy wasm function to Redpanda

> we use `rpk wasm deploy <wasm_definition_file_dist_path>` in this case

Reviewer: ```suggestion
`$ rpk wasm deploy <wasm_definition_file_dist_path>`

Reviewer 2: Related to this: I think it's easier when you just tell people exactly what to do. That is, you create a dir with a specific name, in a specific location, then create a file with a specific name, etc. It will prevent new users from making mistakes or typos, and advanced users will know that they can adjust it to their liking.

```bash
rpk wasm deploy dist/elastic.js
```

> when the coprocessor is loaded we can see the redpanda logs or wasm log

Reviewer: ```suggestion
To verify that the coprocessor has successfully been deployed, you can look through the redpanda logs for a log line that looks like [this]

for check it.

on `redpanda logs`

```bash
INFO [shard 0] coproc - service.cc:101 - Request recieved to register coprocessor with id: 14103244480447969041
```

> on `/var/lib/redpanda/coprocessor/logs/wasm`

Reviewer: ```suggestion
The wasm engine itself also has a logfile which by default will exist in `/var/lib/redpanda/coprocessor/logs/wasm`. To see the status of your transform reported by the wasm engine look for log lines like [this]

```bash
[FileManager] info: Detected new file in submit dir: /var/lib/redpanda/coprocessor/submit/elastic.js

[FileManager] info: Initiating RPC call to redpandas coprocessor service at endpoint - enable_coprocessors with data: [object Object]
TODO: improve [object Object]
```

> ## View result

Reviewer: ```suggestion
Run data through the coprocessor

> to see if our wasm function is working, we need to push information to `produce` topic in Redpanda.

Reviewer: ```suggestion
To transform data, push records onto the `produce` topic using any kafka client.

Author: Done

```bash

> rpk topic produce produce -H logService:elastic -n 5

Reviewer: ```suggestion
In this example we will use the kafka client within `rpk`.
```bash
$ rpk topic produce produce -H logService:elastic -n 5
```

Author: Done

```

> on `/var/lib/redpanda/coprocessor/logs/wasm`

Reviewer: ```suggestion
In `/var/lib/redpanda/coprocessor/logs/wasm` you should see log line that look like [this]

Author: Done

```
Applying wasm function
Index record value to Elastic Search
```

> ## Chech Elastic Search result

Reviewer: ```suggestion
## Verify Elastic Search results

Author: Done

> next code take all records with `index=result_wasm`

Reviewer: ?? confusing no context here

Author: Done I added more context

> # Redpanda wasm function + Elastic Search

Reviewer 3: @andresaristizabal Looks like a great article!
Looks like we need to correct the style on these terms:

- wasm -> Wasm (as shown in https://webassembly.org/)
- ElasticSearch and Elastic Search -> Elasticsearch (as shown in https://www.elastic.co/elasticsearch/)

Do you want me to add that as a suggestion for every instance?

Author: Got it, let me rename all words on this guide.

One of the richest wasm function cases is pushing to [ElasticSeach](https://www.elastic.co/elasticsearch) from a Redpanda topic. To do this, you'll need to satisfy two requirements:

1. Redpanda version 21.1.5 or greater. You can check it by running `rpk version`.
2. ElasticSearch, in this example we use the official [docker image](https://www.elastic.co/guide/en/elasticsearch/reference/current/docker.html)

## Enable wasm on Redpanda
Add `coproc_enable: true` in your redpanda.yaml file, under the `redpanda` section. You can do this manually or by running 

```bash
$ sudo rpk redpanda config set 'redpanda.coproc_enable' true
If redpanda is running, please restart it for the changes to take effect.
```

Also you need to set `developer_mode: true`, for set developer_mode you can use:

```bash
$ sudo rpk redpanda mode dev
Writing 'dev' mode defaults to '/etc/redpanda/redpanda.yaml'
```

expected config:

```yaml
   # Skips most of the checks performed at startup (i.e. memory, xfs)
   # not recomended for production use
   developer_mode: true
   enable_coproc: true # enable wasm
```

## Create Wasm function

> In this example we use a folder for save all wasm functions, `/wasm_example`

Author: Yes, this PR is already done.

First, we need to set up the environment for defining and compacting
wasm function, for it we can use `rpk wasm generate`

```bash
$ rpk wasm generate
npm created project in /wasm_example/wasm
$ /wasm_example tree wasm 
wasm
  ├── package.json
  ├── src
  │   └── wasm.js
  ├── test
  │   └── wasm.test.js
  └── webpack.js
```

that command creates a folder with an example wasm
definition in `/wasm_example/wasm`, now we just to go 
`src/wasm.js`, and rename to `src/elastic.js`, remember
the wasm file name is very important, from this we create
the id for a wasm function. 

```bash
$ /wasm_example tree wasm/src
wasm/src
  └── elastic.js
```

Be sure to add elasticsearch to your install dependencies

```bash
$ npm install @elastic/elasticsearch
```

now we can edit `src/elastic.js` in order to apply our logic

```js
const {
  SimpleTransform,
  PolicyError,
} = require("@vectorizedio/wasm-api");

const {Client} = require('@elastic/elasticsearch')

const client = new Client({ node: 'http://localhost:9200' })

const transform = new SimpleTransform();
/* Topics that fire the transform function */

> transform.subscribe(["produce"]);

Author: Done, I change the topic for `origin`

The statically defined input topic that this wasm operator has defined in its metadata

```bash
rpk topic create produce
```

## Deploy wasm function to Redpanda

we use `rpk wasm deploy ~//wasm_example/wasm/dist/elastic.js` in this case

```bash
$ rpk wasm deploy dist/elastic.js

Sent record to partition 0 at offset 1 with timestamp 2021-02-01 18:18:15.734185538 -0500 -05 m=+0.053943881.
```

To verify that the coprocessor has successfully been deployed, you can look through the redpanda logs for a log line that looks like [this]
for check it.

on `redpanda logs`

```bash
INFO [shard 0] coproc - service.cc:101 - Request recieved to register coprocessor with id: 14103244480447969041
```

The wasm engine itself also has a logfile which by default will exist in `/var/lib/redpanda/coprocessor/logs/wasm`. To see the status of your transform reported by the wasm engine look for log lines like [this]

```bash
[FileManager] info: Detected new file in submit dir: /var/lib/redpanda/coprocessor/submit/elastic.js

[FileManager] info: Initiating RPC call to redpandas coprocessor service at endpoint - enable_coprocessors with data: [object Object]
TODO: improve [object Object]
```

> Run data through the coprocessor

Author: > Should this be a title?

Yes, it is a title

> Run data through the coprocessor

Author: > Also be consistent with terms. Up until now, you've referred to it as "function". People will think a "coprocessor" or "transform" is a diffetent thing.

Done I'll refactor all name with, wasm script and wasm engine, in order to avoid ambiguous names

> Run data through the coprocessor

Reviewer 2: Check with Alex & the sales team first!

To transform data, push records onto the `produce` topic using any kafka client. In this example we will use the kafka client within `rpk`.

```bash

> $ rpk topic produce produce -H logService:elastic -n 5

Reviewer 2: It's worth explaining why the header is needed.

> $ rpk topic produce produce -H logService:elastic -n 5

Author: > It's worth explaining why the header is needed.

Got it, I'll add a note for it

```

In `/var/lib/redpanda/coprocessor/logs/wasm` you should see log line that look like [this]

```
Applying wasm function
Index record value to Elastic Search
```

## Verify Elastic Search results

> Now we need to validate that our wasm function, take a records and publish to ElasticSearch. To do this, we can use the same client for making a query to ElasticSearch, where we want to take all hist that has a wasm index, in this case `result_wasm`.

Reviewer 2: This reads like you're using the following script to publish to ES, when you're actually retrieving results.

```js
const { Client } = require("@elastic/elasticsearch")

const client = new Client({ node: 'http://localhost:9200' })

client.search({
  index: 'result_wasm'
}).then((results) => {
  const results = results.body.hits.hits.map(record => record._source)
  console.log(results)
})
```
 expected output:

 ```bash
 [
  { record_message: 'Information from Redpanda + Wasm' },
  { record_message: 'Information from Redpanda + Wasm' },
  { record_message: 'Information from Redpanda + Wasm' },
  { record_message: 'Information from Redpanda + Wasm' },
  { record_message: 'Information from Redpanda + Wasm' }
]

>  ```

Reviewer 2: The guide needs an end so that the user knows it's finished and they can go on living their lives.

Reviewer 2: Explain why this is the expected output.

Author: Done

/* function for validating record has the correct header key and value  */
const containHeader = (key, value, record) =>
  record.headers.some( header =>
    header.headerKey.equals(Buffer.from(key)) &&
    header.value.equals(Buffer.from(value))
  )

/* Transform function */
transform.processRecord((recordBatch) => {
  console.log("Applying wasm function")
  return Promise.all(
    recordBatch.records.map((record) => {
      if (containHeader("logService", "elastic", record)) {
        console.log("Index record value to Elastic Search")
        return client.index({
          index: "result_wasm",
          type: "redpanda_wasm",
          body: {
            /*
             index the record value into elastic search
            */
            record_message: record.value.toString()
          }
        })
      }
    })
  ).then(() => {
      /*
       return a map with recordBatch indexed to elastic Search
      */
      const result = new Map();
      result.set("elasticIndex", recordBatch)
      /* 
       Wasm processRecord function has to return a Map always,
       although that map is empty map
      */
      return result
  })

>     .catch((e) => {

Reviewer 4: i don't think we should catch this exception. instead we should let it bubble up. and the script deregisters.

Author: Got it. we can remove it.

```bash
$ rpk wasm deploy dist/elastic.js

Sent record to partition 0 at offset 1 with timestamp 2021-02-01 18:18:15.734185538 -0500 -05 m=+0.053943881.
```

To verify that the wasm script has successfully been deployed, 
you can look through the redpanda logs for a log line that looks like this
for check it.

on `redpanda logs`

```bash
INFO [shard 0] coproc - service.cc:101 - Request recieved to register coprocessor with id: 14103244480447969041
```

The Wasm engine itself also has a log file which is located by default in `/var/lib/redpanda/coprocessor/logs/wasm`. To see the status of your function as reported by the Wasm engine look for log lines like [this]

```bash
   2021-03-09T14:33:42.367Z [server] info: request enable wasm script:  14103244480447969041
   2021-03-09T14:33:42.380Z [server] info: wasm script loaded on nodejs engine: 14103244480447969041
```

## Run data through the Wasm engine

To transform data, push records onto the `origin` topic using any 
kafka client. In this example we will use the kafka client 
within `rpk`.

>Note: we need to publish the records with `logService:elastic` 
>headers in order to those records pass the `containHeader` 

> >validation into `processRecord` function on our wasm script.  

Reviewer: `in order for those records to pass the 'containHeader' validator that our wasm script will run for each record`
